### ISSUE TITLE
`InputBag::get()` should be smarter

### ISSUE THREAD

**Issue by vhenzl** [phpstan/phpstan-symfony#177]
While technically correct, the change in 716a70de8524928036d5bc0d880d1117fa77de82 leads to false positive reports for code like this:

```php
$abc = $request->query->get('abc');
Assert::notNull($abc);
expectString($abc); // 💥 Parameter #1 $str of function expectString expects string, bool|float|int|string given.
```

AFAIK Symfony never converts query string values from `string` to other types. So unless the `$defaultValue` argument is specified, `InputBag::get()` always returns `string|null`. And if `$defaultValue` is specified, the return type should be `string|T` where `T` extends `string|int|float|bool|null`.

**Comment by ondrejmirtes** [comment]
Symfony is really confusing here. This change was triggered by https://github.com/symfony/symfony/commit/381a0a19f741da314dff99fc64cc75d674fc7beb - I guess we'd have to make InputBag generic and differentiate between POST and other uses... But I'm not fluent in Symfony enough to know all the places where InputBag can be retrieved from to mark them correctly.

**Comment by jaydiablo** [comment]
We recently upgraded from `0.12.27` to `0.12.38` of this plugin and started getting a lot of false positives on code like this:

`$request->request->get('field_name', '');`

`$request->request` is an InputBag, and the previous behavior seemed that it would use the second parameter of the `get` method to determine what the type would be (in this case, a `string`), now it appears that it sees it as `bool|float|int|string`.

**Comment by benr77** [comment]
I agree - it should look at the second parameter to `get()` which is the default value and determine the expected type from that. If it's not specified, it should fall back to `string|null`

**Comment by ondrejmirtes** [comment]
@benr77 I disagree. You can have `$inputBag->get('foo', 1)` and it can still return `string`.

**Comment by benr77** [comment]
@ondrejmirtes OK good point you are correct. 

The problem remains though - how do we get around all these false positives? Is there a workaround or is just ignoring them for the moment the best solution? Thanks

**Comment by ondrejmirtes** [comment]
See https://github.com/phpstan/phpstan-symfony/issues/177#issuecomment-872005004 - if someones compiles a list of all possible places to get `InputBag` in a typical Symfony application, we can differentiate between `POST` and other usages.

**Comment by benr77** [comment]
Well `InputBag` is only used for POST/GET/COOKIE parameters, so certainly for any correctly designed app this should only be used in controller classes. I have checked several of my Symfony apps and the only place I see `new InputBag()` is in the `http-foundation/Request.php`

In the 5.1.0 changelog there is:

`Deprecated retrieving non-string values using `InputBag::get()`, use `InputBag::all()` if you need access to the collection of values`

Based on this, can we assume `string|null` is the expected return type of `InputBag::get()` ??

**Comment by ondrejmirtes** [comment]
This commit shows that the intended message is "non-scalar" values, not "non-string" values: https://github.com/symfony/symfony/commit/381a0a19f741da314dff99fc64cc75d674fc7beb

I'm asking how a `GET` `InputBag` is retrieved as opposed to `POST` `InputBag` for example...

**Comment by benr77** [comment]
Do you mean:

```
/** @var Request $request */

// POST vars InputBag
$request->request

// GET vars InputBag
$request->query

//COOKIES vars InputBag
$request->cookies
```

https://symfony.com/doc/current/components/http_foundation.html#accessing-request-data

**Comment by ondrejmirtes** [comment]
Two more questions:

1) Any other way to obtain an InputBag?
2) Is it usual for users to typehint InputBag in their application somewhere?

**Comment by benr77** [comment]
1) Not to my knowledge - but I suppose you could leverage it directly yourself for some reason, but why this would be useful over just using `ParameterBag` directly I don't know.
2) No, I don't think this would be very likely.

**Comment by ondrejmirtes** [comment]
Alright, so my plan is to introduce some stubs that make InputBag generic, and then to typehint those Request properties properly.

**Comment by vhenzl** [comment]
Generic `InputBag` should be fine for `$request->query` and `$request->cookies` but I'm not sure about `$request->request`. The `request` field can hold decoded JSON payload. It's mentioned in https://github.com/symfony/symfony/pull/40315#issuecomment-786609517 and is the reason why the type-hint in `InputBag::get()` has been updated in the first place. 

Anyone could theoretically replace their `query` and `cookies` as well but I can't see any reason why anyone would do that.

It'd done by replacing the (until then empty) `request` by decoded content; basically  `$request->request->replace(json_decode($request->getContent(), true))`.

Symfony doesn't have any built-in mechanism to do so, it's up to the developer to do it manually or use packages like https://github.com/symfony-bundles/json-request-bundle or https://github.com/qandidate-labs/symfony-json-request-transformer.

There is also `Request::toArray()` that has been introduced recently (Symfony 5.2) and is meant for obtaining parsed JSON, but I reckon that majority still uses `request` for it (which IMO is a more correct approach).

**Comment by ondrejmirtes** [comment]
@vhenzl So is this fine or incorrect? What would you do?

```php
	/**
	 * Request body parameters ($_POST).
	 *
	 * @var InputBag<string|int|float|bool>
	 */
	public $request;
```

**Comment by ondrejmirtes** [comment]
Please test this before I release it: https://github.com/phpstan/phpstan-symfony/commit/95d9ae78605e5023e0edcf905c9b42da2a2a8359 (you'll need phpstan/phpstan dev-master too)

**Comment by vhenzl** [comment]
> So is this fine or incorrect?

Fine. It should be a sensible default for most people I believe.

I have just realised that `InputBag:::get()` can't return `array` ([on purpose](https://github.com/symfony/symfony/pull/41766)) so it makes it kinda unusable for parsed JSON. But if you have a parsed JSON payload in `$request->request`, you are unlikely to access it via `get()` anyway, you would just get the whole thing via `$request->request->all()`.

So the use case for `request->get()` is probably only with form submissions. And in that case, the return is likely to be `string|null` only, but having `string|int|float|bool|null` is safer. 

Having `string|null` for `query` and `cookies` should be perfectly fine as their content comes directly from `$_GET` and `$_COOKIE`.

**Comment by vhenzl** [comment]
Thanks, @ondrejmirtes! 

I tried it and it works as expected for me. Needless to say, though, that we have just a few places with `get()`, the majority of our code uses `all()`, so @jaydiablo and @benr77 are better to give it try as well.

**Comment by benr77** [comment]
I just tested it too and it seems to be working well. I would be happy for you to release. Thanks for your efforts!

**Comment by ondrejmirtes** [comment]
Awesome :) I'm gonna release it along with PHPStan 0.12.97 in the coming days.

**Comment by Tobion** [comment]
The solution does not seem right to me. See https://github.com/phpstan/phpstan-symfony/commit/95d9ae78605e5023e0edcf905c9b42da2a2a8359#r56194881
I get alot of false positive phpstan errors like

> Call to function is_string() with string will always evaluate to true.

**Comment by ondrejmirtes** [comment]
@Tobion When do you think it doesn't return a string?

**Comment by Tobion** [comment]
The setter accepts any scalar. So the getter can also return any scalar.

**Comment by ondrejmirtes** [comment]
@Tobion Do you have a use-case where you're setting values to an InputBag that you've got from Request through one of the public properties? Please open a new issue about that.

**Comment by Tobion** [comment]
I don't know. But if it doesn't make sense, it shouldn't be like this in Symfony. It was changed recently to better support static analysis, but apparently it didn't make it better then if there are still workarounds needed.
